scipy/stats/stats.py

```diff
@@ -3294,26 +3294,29 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate'):
     """
     a, axisout = _chk_asarray(a, axis)
 
-    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+    a_contains_nan, nan_policy = _contains_nan(a, nan_policy)
 
-    if contains_nan and nan_policy == 'omit':
+    if a_contains_nan and nan_policy == 'omit':
         a = ma.masked_invalid(a)
-        b = ma.masked_invalid(b)
-        return mstats_basic.spearmanr(a, b, axis)
 
-    if a.size <= 1:
+    if a.size <= 1 or (a_contains_nan and nan_policy == 'propagate'):
         return SpearmanrResult(np.nan, np.nan)
+        
     ar = np.apply_along_axis(rankdata, axisout, a)
 
     br = None
     if b is not None:
         b, axisout = _chk_asarray(b, axis)
 
-        contains_nan, nan_policy = _contains_nan(b, nan_policy)
+        b_contains_nan, nan_policy = _contains_nan(b, nan_policy)
+
+        if a_contains_nan or b_contains_nan:
+            if nan_policy == 'propagate':
+                return SpearmanrResult(np.nan, np.nan)
```

Comment:
I added some changes that provide n-dim array of np.nan with `propagate` option. In the last change i evaluate correlation and pvaluy no matter of `nan_policy` option and after that i check `nan_policy` option. 

Better solution will be calculation size of output array without calculation of correlation and pvalue but i don't know how to calculate this size (i have idea to use sum a.shape[1] and b.shape[1] but i don't sure in that solution).